doc/performance-testing/cloud-run/0001_stress-test-december-2021.md

```diff
@@ -0,0 +1,86 @@
+
+- The 99th percentile max cpu usage as reported in Grafana
+
+## Errors and Failures
+
+- 60 Clients per instance:
+    - 4 x "Expected a (200) but got a (0) back when getting page"
+- 70 Clients per instance:
+    - 4 x "Expected a (200) but got a (0) back when getting page"
+    - 1 x "Expected a (302) but got a (0) back when getting page"
+- 150 Clients per instance:
+  - 1 x "Expected a (302) but got a (0) back when getting page"
+- 160 Clients per instance:
+  - 3 x "Expected a (200) but got a (0) back when getting page"  
+
+## Observations
+
+- Some early errors that didn't persist when we increased the load.
+- The following CPU usage thresholds were exceeded:
+  - 50% at 80 clients per instance.
+  - 60% at 100 clients per instance.
+  - 70% at 120 clients per instance.
+  - 80% CPU at 130 clients per instance.
+  - 90% CPU at 160 clients per instance.
+- Significant performance degradation on 99th percentile response time between 100 and 110 clients per instance.
+- We reached over 300ms response on 99% at 120 clients per instance.
+- Above 130 clients per instance CPU levels off because response times are degrading significantly.
+- Even at 60-75% CPU, the response time is still performant.
+- Above 150 clients per instance performance degradation is severe.
+- Over the course of the test there were 42k survey submissions.
+
+#Additional tests
+Following our initial stress test we decided to run an additional test to prove that performance stays on similar levels when we fine tune runner's concurrency. Based on the table above we made an assumption that if we set concurrency to 90 we should get similar results. In addition, if we update number of min. max instances to 2 and double the number of clients per instance we should also obtain similar results.
+
+##Comparison
+
+| Clients per instance | Concurrency | Min. Max. instances | Requests per second | 99th percentile Max CPU Usage (%) | 99th percentile response time (ms) | Error rate (%) | Output |
+| --- | --- | --- | --- | --- | --- | --- | --- |
+| 90  | 250 | 1 | 106.09 | 56.0 | 170 |  0.000 | [output](https://console.cloud.google.com/storage/browser/eq-stress-injector-07122021-outputs/stress-test/2021-12-08T14:42:00) |
+| 90  | 90 | 2 | 105.74 | 38.0 | 160 | 0.000 | [output](https://console.cloud.google.com/storage/browser/eq-stress-injector-07122021-outputs/stress-test/2021-12-09T14:30:33) |
+| 90  | 90 | 1 | 104.21 | 66.0 | 190 | 0.001 | [output](https://console.cloud.google.com/storage/browser/eq-stress-injector-07122021-outputs/stress-test/2021-12-09T13:39:52) |
+
+
+## Observations
+
+- When we fine-tuned runner's concurrency to be 90 and only one instance, the performance was slightly worse than we would expect (66% CPU usage compared to 56%). However, when we used 90 with 2 instances the performance significantly improved (e.g. CPU usage decreased and latency improved).
```

Comment:
Was it repeated to check if it was an anomly?. Also, this could be because the concurrency limit is a hard limit, any requests above it is returned as a 429 if we reached max scaling. Therefore, when we come to use Cloud Run in prod, we will need to set max instances to much higher than when runner used to be under k8s. GCP also uses CPU utilisation to scale, and it tries to keep it at [60% avg according to the on doc](https://cloud.google.com/run/docs/about-instance-autoscaling), therefore, we might be better off setting the concurrency slightly higher at like 100/110.

I think it is also worth noting above that the CPU utilisation metric for Cloud Run is measured across all containers, so 38% CPU on 2 instances could mean one container was 50%, and the other container was 26%.

Reply:
This will be investigated in the next phase of testing which will be specifically around scaling